Source/WebKit/WPEPlatform/wpe/WPEGestureController.cpp

```diff
@@ -0,0 +1,269 @@
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+ * PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "WPEGestureController.h"
+
+#include <cmath>
+#include <optional>
+#include <wtf/glib/WTFGType.h>
+
+namespace WPE {
+
+class GestureDetector final {
+public:
+    WPEGesture detectGesture(WPEEvent* event)
+    {
+        WPEGesture detectedGesture = WPE_GESTURE_NONE;
+
+        switch (wpe_event_get_event_type(event)) {
+        case WPE_EVENT_TOUCH_DOWN:
+            if (double x, y; wpe_event_get_position(event, &x, &y))
+                m_touchDownPoint = Point { x, y };
+            break;
+        case WPE_EVENT_TOUCH_CANCEL:
+            reset();
+            break;
+        case WPE_EVENT_TOUCH_MOVE:
+            if (!m_touchDownPoint)
+                break;
+            if (double x, y; wpe_event_get_position(event, &x, &y)) {
+                if (!m_dragReferencePoint && std::hypot(x - m_touchDownPoint->x, y - m_touchDownPoint->y) > dragActivationThresholdPx)
+                    m_dragReferencePoint = m_touchDownPoint;
+                if (m_dragReferencePoint) {
+                    detectedGesture = WPE_GESTURE_DRAG;
+                    m_detectedGesturePoint = m_touchDownPoint;
+                    m_detectedGestureDelta = { x - m_dragReferencePoint->x, y - m_dragReferencePoint->y };
+                    m_dragReferencePoint = Point { x, y };
+                }
+            }
+            break;
+        case WPE_EVENT_TOUCH_UP:
+            if (double x, y; wpe_event_get_position(event, &x, &y) && m_touchDownPoint) {
+                if (m_dragReferencePoint) {
+                    detectedGesture = WPE_GESTURE_DRAG;
+                    m_detectedGesturePoint = m_touchDownPoint;
+                    m_detectedGestureDelta = { x - m_dragReferencePoint->x, y - m_dragReferencePoint->y };
+                } else if (m_touchDownPoint) {
+                    detectedGesture = WPE_GESTURE_TAP;
+                    m_detectedGesturePoint = m_touchDownPoint;
+                    m_detectedGestureDelta = std::nullopt;
+                }
+            }
+            reset();
```

Comment:
I think it would be simpler if we reset in the next down, so that after UP we can query the gesture controller.